sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultServiceSpec.java

```diff
@@ -370,6 +380,38 @@ public ServiceSpec parse(byte[] bytes) throws ConfigStoreException {
         public static final Collection<Class<?>> getDefaultRegisteredSubtypes() {
             return defaultRegisteredSubtypes;
         }
+
+        /**
+         * Custom deserializer for goal states to accomodate transition from FINISHED to ONCE/FINISH.
+         */
+        public static class GoalStateDeserializer extends StdDeserializer<GoalState> {
+
+            public GoalStateDeserializer() {
+                this(null);
+            }
+
+            protected GoalStateDeserializer(Class<?> vc) {
+                super(vc);
+            }
+
+            @Override
+            public GoalState deserialize(
+                    JsonParser p, DeserializationContext ctxt) throws IOException, JsonParseException {
+                String value = ((TextNode) p.getCodec().readTree(p)).textValue();
+
+                if (value.equals("FINISHED") || value.equals("ONCE")) {
+                    return GoalState.ONCE;
+                } else if (value.equals("FINISH")) {
+                    return GoalState.FINISH;
+                } else if (value.equals("RUNNING")) {
+                    return GoalState.RUNNING;
+                } else if (value.equals("UNKNOWN")) {
+                    return GoalState.UNKNOWN;
+                }
+
+                throw new JsonParseException(String.format("Unknown goal state: %s", value), p.getCurrentLocation());
```

Comment:
What do the logs look like when this is thrown?

Reply:
Very similar to a Python syntax error, e.g. ```JsonParseException, line 1, column 37: Unknown goal state: STATE```, followed by the usual stack trace.

Reply:
So no scheduler crash?

Reply:
Oh, it'll crash, since it ultimately results in an uncaught `ConfigStoreException` that will break during scheduler building.